drivers/counter/counter_nxp_pit.c

```diff
@@ -0,0 +1,407 @@
+/*
+ * Copyright 2020-24 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_pit
+
+#include <zephyr/drivers/counter.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/irq.h>
+#include <fsl_pit.h>
+
+#define LOG_MODULE_NAME counter_pit
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Device holds a pointer to pointer to data */
+#define PIT_CHANNEL_DATA(dev)					\
+	(*(struct nxp_pit_channel_data *const *const)dev->data)
+
+/* Device config->data is an array of data pointers ordered by channel number,
+ * dev->data is a pointer to one of these pointers in that array,
+ * so the value of the dev->data - dev->config->data is the channel index
+ */
+#define PIT_CHANNEL_ID(dev)					\
+	(((struct nxp_pit_channel_data *const *)dev->data) -	\
+	((const struct nxp_pit_config *)dev->config)->data)
+
+
+struct nxp_pit_channel_data {
+	uint32_t top;
+	counter_top_callback_t top_callback;
+	void *top_user_data;
+};
+
+
+struct nxp_pit_config {
+	struct counter_config_info info;
+	PIT_Type *base;
+	bool enableRunInDebug;
+	int num_channels;
+#if DT_NODE_HAS_PROP(DT_COMPAT_GET_ANY_STATUS_OKAY(nxp_pit), interrupts)
+	void (*irq_config_func)(const struct device *dev);
+#else
+	void (**irq_config_func)(const struct device *dev);
+#endif
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	struct nxp_pit_channel_data *const *data;
+	const struct device *const *channels;
+};
+
+static uint32_t nxp_pit_get_top_value(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	pit_chnl_t channel = PIT_CHANNEL_ID(dev);
+
+	/*
+	 * According to RM, the LDVAL trigger = clock ticks -1
+	 * The underlying HAL driver function PIT_SetTimerPeriod()
+	 * automatically subtracted 1 from the value that ends up in
+	 * LDVAL so for reporting purposes we need to add it back in
+	 * here to by consistent.
+	 */
+	return (config->base->CHANNEL[channel].LDVAL + 1);
+}
+
+static int nxp_pit_start(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	int channel_id = PIT_CHANNEL_ID(dev);
+
+	LOG_DBG("period is %d", nxp_pit_get_top_value(dev));
+	PIT_EnableInterrupts(config->base, channel_id,
+			     kPIT_TimerInterruptEnable);
+	PIT_StartTimer(config->base, channel_id);
+	return 0;
+}
+
+static int nxp_pit_stop(const struct device *dev)
+{
+	const struct nxp_pit_config *config = dev->config;
+	int channel_id = PIT_CHANNEL_ID(dev);
+
+	PIT_DisableInterrupts(config->base, channel_id,
```

Comment:
what's the need to enable/disable interrupts in start/stop functions?

Reply:
also I guess the SDK handles the difference between when theres one interrupt vs many?

Reply:
At this point in the code flow, we are dealing with individual devices, a device per channel, I imagine if you have the ability, you would want to ensure interrupts are available as the timer is active and disabled when the timer is no longer of use to the user, there may be more potential to look at in regards to these interrupts but might be a subject for a different PR.